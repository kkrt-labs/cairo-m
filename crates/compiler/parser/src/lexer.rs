// crates/cairo-m-compiler-parser/src/tokens.rs
use std::fmt;

use logos::Logos;

/// Custom error type for lexing errors in the Cairo-M language.
///
/// This enum represents different types of errors that can occur during lexical analysis,
/// providing detailed information about what went wrong and where.
#[derive(Default, Debug, Clone, PartialEq, Eq)]
pub enum LexingError {
    /// An integer literal that cannot be parsed or is out of range
    InvalidNumber {
        /// The problematic number string that failed to parse
        value: String,
        /// The reason for the parsing failure
        reason: NumberParseError,
    },
    /// A character that is not recognized as part of any valid token
    #[default]
    InvalidCharacter,
}

/// Specific errors that can occur when parsing numeric literals.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum NumberParseError {
    /// The number is too large to fit in the target integer type (u32)
    Overflow,
    /// The number format is invalid (e.g., invalid characters for the base)
    InvalidFormat,
    /// An unknown parsing error occurred
    Unknown(String),
}

impl fmt::Display for LexingError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::InvalidNumber { value, reason } => {
                write!(f, "Invalid number '{value}': {reason}")
            }
            Self::InvalidCharacter => {
                write!(f, "Invalid character")
            }
        }
    }
}

impl fmt::Display for NumberParseError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Overflow => write!(f, "Value is higher than u32::max"),
            Self::InvalidFormat => write!(f, "Invalid number format"),
            Self::Unknown(s) => write!(f, "{s}"),
        }
    }
}

/// Helper function to parse numeric literals with proper error reporting.
///
/// This function handles different numeric bases (decimal, hexadecimal, octal, binary)
/// and provides detailed error information when parsing fails.
fn parse_number_literal<'a>(lex: &logos::Lexer<'a, TokenType<'a>>) -> Result<u32, LexingError> {
    let slice = lex.slice();

    // Parse based on prefix to determine the base
    let (number_str, base) = if slice.starts_with("0x") || slice.starts_with("0X") {
        (&slice[2..], 16)
    } else if slice.starts_with("0o") || slice.starts_with("0O") {
        (&slice[2..], 8)
    } else if slice.starts_with("0b") || slice.starts_with("0B") {
        (&slice[2..], 2)
    } else {
        (slice, 10)
    };

    // Parse the number string as u32
    match u32::from_str_radix(number_str, base) {
        Ok(n) => Ok(n),
        Err(err) => {
            let reason = match err.kind() {
                std::num::IntErrorKind::PosOverflow | std::num::IntErrorKind::NegOverflow => {
                    NumberParseError::Overflow
                }
                std::num::IntErrorKind::InvalidDigit => NumberParseError::InvalidFormat,
                _ => NumberParseError::Unknown(err.to_string()),
            };

            Err(LexingError::InvalidNumber {
                value: slice.to_string(),
                reason,
            })
        }
    }
}

#[derive(Logos, Debug, Clone, Copy, PartialEq, Eq)]
#[logos(skip r"[\t\n\r ]+")] // Skip whitespace, including carriage return
#[logos(skip r"//[^\n]*")] // Skip single-line comments
#[logos(error = LexingError)] // Use our custom error type
pub enum TokenType<'a> {
    // Literals
    #[regex(
        r"[0-9]+|0[xX][0-9a-fA-F]*[a-zA-Z]*|0[oO][0-7]*[a-zA-Z0-9]*|0[bB][01]*[a-zA-Z0-9]*",
        parse_number_literal
    )]
    LiteralNumber(u32),
    // Keywords
    #[token("as")]
    As,
    #[token("const")]
    Const,
    #[token("else")]
    Else,
    #[token("false")]
    False,
    #[token("fn")]
    Function,
    #[token("if")]
    If,
    #[token("let")]
    Let,
    #[token("namespace")]
    Namespace,
    #[token("return")]
    Return,
    #[token("struct")]
    Struct,
    #[token("true")]
    True,
    #[token("while")]
    While,
    #[token("loop")]
    Loop,
    #[token("for")]
    For,
    #[token("in")]
    In,
    #[token("break")]
    Break,
    #[token("continue")]
    Continue,
    #[token("use")]
    Use,
    // Identifiers (must come after keywords)
    #[regex(r"[a-zA-Z_][a-zA-Z0-9_]*")]
    Identifier(&'a str),
    // Operators (order matters for longest match)
    #[token("!")]
    Not,
    #[token("&&")]
    AndAnd,
    #[token("||")]
    OrOr,
    #[token("==")]
    EqEq,
    #[token("!=")]
    Neq,
    #[token("<=")]
    LessEqual,
    #[token(">=")]
    GreaterEqual,
    #[token("<")]
    Less,
    #[token(">")]
    Greater,
    #[token("+")]
    Plus,
    #[token("-")]
    Minus,
    #[token("*")]
    Mul,
    #[token("/")]
    Div,
    #[token("=")]
    Eq,
    // Punctuation
    #[token("->")]
    Arrow,
    #[token("(")]
    LParen,
    #[token(")")]
    RParen,
    #[token("{")]
    LBrace,
    #[token("}")]
    RBrace,
    #[token("[")]
    LBrack,
    #[token("]")]
    RBrack,
    #[token(",")]
    Comma,
    #[token(";")]
    Semicolon,
    #[token(":")]
    Colon,
    #[token("::")]
    ColonColon,
    #[token(".")]
    Dot,
}

impl<'a> fmt::Display for TokenType<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            TokenType::LiteralNumber(n) => write!(f, "{n}"),
            TokenType::Identifier(s) => write!(f, "{s}"),
            TokenType::As => write!(f, "as"),
            TokenType::Const => write!(f, "const"),
            TokenType::Else => write!(f, "else"),
            TokenType::False => write!(f, "false"),
            TokenType::Function => write!(f, "fn"),
            TokenType::If => write!(f, "if"),
            TokenType::Let => write!(f, "let"),
            TokenType::Namespace => write!(f, "namespace"),
            TokenType::Return => write!(f, "return"),
            TokenType::Struct => write!(f, "struct"),
            TokenType::True => write!(f, "true"),
            TokenType::While => write!(f, "while"),
            TokenType::Loop => write!(f, "loop"),
            TokenType::For => write!(f, "for"),
            TokenType::In => write!(f, "in"),
            TokenType::Break => write!(f, "break"),
            TokenType::Continue => write!(f, "continue"),
            TokenType::Not => write!(f, "!"),
            TokenType::AndAnd => write!(f, "&&"),
            TokenType::OrOr => write!(f, "||"),
            TokenType::EqEq => write!(f, "=="),
            TokenType::Neq => write!(f, "!="),
            TokenType::LessEqual => write!(f, "<="),
            TokenType::GreaterEqual => write!(f, ">="),
            TokenType::Less => write!(f, "<"),
            TokenType::Greater => write!(f, ">"),
            TokenType::Plus => write!(f, "+"),
            TokenType::Minus => write!(f, "-"),
            TokenType::Mul => write!(f, "*"),
            TokenType::Div => write!(f, "/"),
            TokenType::Eq => write!(f, "="),
            TokenType::Arrow => write!(f, "->"),
            TokenType::LParen => write!(f, "("),
            TokenType::RParen => write!(f, ")"),
            TokenType::LBrace => write!(f, "{{"),
            TokenType::RBrace => write!(f, "}}"),
            TokenType::LBrack => write!(f, "["),
            TokenType::RBrack => write!(f, "]"),
            TokenType::Comma => write!(f, ","),
            TokenType::Semicolon => write!(f, ";"),
            TokenType::Colon => write!(f, ":"),
            TokenType::ColonColon => write!(f, "::"),
            TokenType::Dot => write!(f, "."),
            TokenType::Use => write!(f, "use"),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_basic_lexer() {
        let input = r#"
            use std::math::add;

            fn add(x: felt, y: felt) -> felt {
                let result = x + y;
                if result == 0 {
                    return result;
                } else {
                    return 0;
                }
            }

            let value = add(10, 20);
            const MAX_SIZE = 100;
            let array = alloc();
            array[1];
        "#;
        let lexer = TokenType::lexer(input);

        let mut tokens = vec![];
        let mut errors = vec![];
        for (token, span) in lexer.spanned() {
            match token {
                Ok(token) => tokens.push(token),
                Err(e) => {
                    errors.push((span, e));
                }
            }
        }

        if !errors.is_empty() {
            panic!("lexer errors: {errors:?}");
        }

        let expected = vec![
            TokenType::Use,
            TokenType::Identifier("std"),
            TokenType::ColonColon,
            TokenType::Identifier("math"),
            TokenType::ColonColon,
            TokenType::Identifier("add"),
            TokenType::Semicolon,
            TokenType::Function,
            TokenType::Identifier("add"),
            TokenType::LParen,
            TokenType::Identifier("x"),
            TokenType::Colon,
            TokenType::Identifier("felt"),
            TokenType::Comma,
            TokenType::Identifier("y"),
            TokenType::Colon,
            TokenType::Identifier("felt"),
            TokenType::RParen,
            TokenType::Arrow,
            TokenType::Identifier("felt"),
            TokenType::LBrace,
            TokenType::Let,
            TokenType::Identifier("result"),
            TokenType::Eq,
            TokenType::Identifier("x"),
            TokenType::Plus,
            TokenType::Identifier("y"),
            TokenType::Semicolon,
            TokenType::If,
            TokenType::Identifier("result"),
            TokenType::EqEq,
            TokenType::LiteralNumber(0),
            TokenType::LBrace,
            TokenType::Return,
            TokenType::Identifier("result"),
            TokenType::Semicolon,
            TokenType::RBrace,
            TokenType::Else,
            TokenType::LBrace,
            TokenType::Return,
            TokenType::LiteralNumber(0),
            TokenType::Semicolon,
            TokenType::RBrace,
            TokenType::RBrace,
            TokenType::Let,
            TokenType::Identifier("value"),
            TokenType::Eq,
            TokenType::Identifier("add"),
            TokenType::LParen,
            TokenType::LiteralNumber(10),
            TokenType::Comma,
            TokenType::LiteralNumber(20),
            TokenType::RParen,
            TokenType::Semicolon,
            TokenType::Const,
            TokenType::Identifier("MAX_SIZE"),
            TokenType::Eq,
            TokenType::LiteralNumber(100),
            TokenType::Semicolon,
            TokenType::Let,
            TokenType::Identifier("array"),
            TokenType::Eq,
            TokenType::Identifier("alloc"),
            TokenType::LParen,
            TokenType::RParen,
            TokenType::Semicolon,
            TokenType::Identifier("array"),
            TokenType::LBrack,
            TokenType::LiteralNumber(1),
            TokenType::RBrack,
            TokenType::Semicolon,
        ];

        assert_eq!(tokens, expected);
    }

    #[test]
    fn test_should_err_on_number_too_large() {
        let u32_max = format!("{}", (u32::MAX as u64 + 1));
        let input = format!("let x = {};", u32_max);
        let lexer = TokenType::lexer(&input);
        let tokens = lexer.spanned().collect::<Vec<_>>();
        assert_eq!(tokens.len(), 5);
        assert_eq!(tokens[0].0, Ok(TokenType::Let));
        assert_eq!(tokens[1].0, Ok(TokenType::Identifier("x")));
        assert_eq!(tokens[2].0, Ok(TokenType::Eq));

        // Check that we get a proper LexingError for the oversized number
        match &tokens[3].0 {
            Err(LexingError::InvalidNumber { value, reason }) => {
                assert_eq!(value, &u32_max);
                assert_eq!(reason, &NumberParseError::Overflow);
            }
            _ => panic!(
                "Expected InvalidNumber error for oversized integer, got: {:?}",
                tokens[3].0
            ),
        }

        assert_eq!(tokens[4].0, Ok(TokenType::Semicolon));
    }

    #[test]
    fn test_invalid_number_formats() {
        // Test invalid hexadecimal
        let input = "0xGG";
        let lexer = TokenType::lexer(input);
        let tokens: Vec<_> = lexer.spanned().collect();
        assert_eq!(tokens.len(), 1);
        match &tokens[0].0 {
            Err(LexingError::InvalidNumber { value, reason }) => {
                assert_eq!(value, "0xGG");
                assert_eq!(reason, &NumberParseError::InvalidFormat);
            }
            _ => panic!(
                "Expected InvalidNumber error for invalid hex, got: {:?}",
                tokens[0].0
            ),
        }

        // Test invalid binary
        let input = "0b123";
        let lexer = TokenType::lexer(input);
        let tokens: Vec<_> = lexer.spanned().collect();
        assert_eq!(tokens.len(), 1);
        match &tokens[0].0 {
            Err(LexingError::InvalidNumber { value, reason }) => {
                assert_eq!(value, "0b123");
                assert_eq!(reason, &NumberParseError::InvalidFormat);
            }
            _ => panic!(
                "Expected InvalidNumber error for invalid binary, got: {:?}",
                tokens[0].0
            ),
        }

        // Test invalid octal
        let input = "0o89";
        let lexer = TokenType::lexer(input);
        let tokens: Vec<_> = lexer.spanned().collect();
        assert_eq!(tokens.len(), 1);
        match &tokens[0].0 {
            Err(LexingError::InvalidNumber { value, reason }) => {
                assert_eq!(value, "0o89");
                assert_eq!(reason, &NumberParseError::InvalidFormat);
            }
            _ => panic!(
                "Expected InvalidNumber error for invalid octal, got: {:?}",
                tokens[0].0
            ),
        }
    }

    #[test]
    fn test_valid_number_formats() {
        // Test all valid number formats
        let test_cases = vec![
            ("42", 42),
            ("0x2A", 42),
            ("0o52", 42),
            ("0b101010", 42),
            ("0", 0),
            ("0x0", 0),
            ("0o0", 0),
            ("0b0", 0),
            ("2147483647", 2147483647), // Max value (2^31 - 1)
        ];

        for (input, expected) in test_cases {
            let lexer = TokenType::lexer(input);
            let tokens: Vec<_> = lexer.spanned().collect();
            assert_eq!(tokens.len(), 1, "Input: {input}");
            match &tokens[0].0 {
                Ok(TokenType::LiteralNumber(n)) => {
                    assert_eq!(*n, expected, "Input: {input}");
                }
                other => panic!("Expected LiteralNumber for input '{input}', got: {other:?}"),
            }
        }
    }
}
