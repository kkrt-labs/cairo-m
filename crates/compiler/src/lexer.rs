use ariadne::{Color, Label, Report, ReportKind, Source};
use logos::Logos;

#[derive(Debug, Clone)]
pub struct Token {
    pub token_type: TokenType,
    pub lexeme: String,
    pub span: (usize, usize),
}


#[derive(Logos, Debug, Clone, PartialEq, Eq)]
#[logos(skip r"[\t\n ]+")]
#[logos(skip r"//.*\n")]
pub enum TokenType {

    #[regex("[0-9]+")]
    Int,

    #[regex(r"%\{(.*)%\}")]
    Hint,

    #[regex(r"0x[0-9a-fA-F]+")]
    HexInt,
    #[regex(r"[a-zA-Z_][a-zA-Z_0-9]*(\.[a-zA-Z_][a-zA-Z_0-9]*)*")]
    Identifier,
    #[regex(r#"".""#)]
    String,
    #[regex(r"'.'")]
    ShortString,

    #[token("++")]
    PlusPlus,
    #[token("==")]
    DoubleEq,
    #[token("**")]
    DoubleStar,
    #[token("!=")]
    Neq,
    #[token("->")]
    Arrow,
    #[token("@")]
    At,

    #[token(",")]
    Comma,

    #[token(":")]
    Colon,

    #[token("felt")]
    Felt,
    #[token("codeoffset")]
    CodeOffset,

    #[token("*")]
    Star,
    #[token("(")]
    LParen,
    #[token(")")]
    RParen,
    
    #[token("=")]
    Equal,

    #[token("+")]
    Plus,
    #[token("-")]
    Minus,
    #[token("/")]
    Slash,

    #[token("&")]
    Ampersand,
    #[token("new")]
    New,

    #[token(".")]
    Dot,

    //Atom
    #[token("nondet")]
    NonDet,
    #[token("[")]
    LBracket,
    #[token("]")]
    RBracket,
    #[token("cast")]
    Cast,

    //Reg
    #[token("ap")]
    Ap,
    #[token("fp")]
    Fp,

    #[token("and")]
    And,
    

    #[token("local")]
    Local,

    #[token("ret")]
    Ret,

    //Instructions

    #[token("call")]
    Call,
    #[token("rel")]
    Rel,
    #[token("abs")]
    Abs,
    #[token("jmp")]
    Jmp,
    #[token("if")]
    If,
    #[token("+=")]
    PlusEq,
    #[token("dw")]
    Dw,
    #[token("{")]
    LBrace,
    #[token("}")]
    RBrace,

    // Import statement
    #[token("import")]
    Import,
    #[token("from")]
    From,
    #[token("as")]
    As,

    // Function/Namespace/Struct definition.
    #[token("func")]
    Func,
    #[token("with")]
    With,
    #[token("struct")]
    Struct,
    #[token("namespace")]
    Namespace,
    #[token("with_attr")]
    WithAttr,
    #[token("else")]
    Else,

    // Cairo file
    // #[token("\n")]
    // Newline,


    // Code elements
    #[token(";")]
    Semicolon,
    #[token("const")]
    Const,
    #[token("let")]
    Let,
    #[token("tempvar")]
    TempVar,
    #[token("assert")]
    Assert,
    #[token("static_assert")]
    StaticAssert,
    #[token("return")]
    Return,
    #[token("using")]
    Using,
    #[token("alloc_locals")]
    AllocLocals,

    EOF,
    Error,

}


pub fn lex(input: &str, file_name: &str) -> (Vec<Token>, u32) {
    let mut error_counter = 0;
    let mut lex = TokenType::lexer(input);
    let mut tokens = Vec::new();
    while let Some(token) = lex.next() {
        let lexeme = lex.slice().to_string();
        if let Ok(token) = token {
            let lexeme = lex.slice().to_string();
            tokens.push(Token {
                token_type: token,
                lexeme,
                span: (lex.span().start, lex.span().end),
            });
        } else {
            let error_span = (file_name, lex.span().start..lex.span().end);
            let _ = Report::build(ReportKind::Error, error_span.clone())
                .with_message("Lexer error")
                .with_label(Label::new(error_span)
                    .with_message(format!("Unknown token '{}'", lexeme))
                    .with_color(Color::Red))
                .finish()
                .print((file_name, Source::from(input)));
            error_counter += 1;
        }
    }
    tokens.push(Token {
        token_type: TokenType::EOF,
        lexeme: "".to_string(),
        span: (0, 0),
    });
    (tokens, error_counter)
}