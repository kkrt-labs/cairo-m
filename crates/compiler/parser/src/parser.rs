use crate::lexer::TokenType;
use chumsky::{input::ValueInput, prelude::*};

#[derive(Debug, Clone)]
#[allow(dead_code)]
enum Statement {
    // TODO: add actual statements here
    Null,
}

#[allow(dead_code)]
fn parser<'tokens, 'src: 'tokens, I>(
) -> impl Parser<'tokens, I, Statement, extra::Err<Rich<'tokens, TokenType<'src>>>>
where
    I: ValueInput<'tokens, Token = TokenType<'src>, Span = SimpleSpan>,
{
    // BOILERPLATE code - TODO
    recursive(|_statement| just(TokenType::Let).map(|_| Statement::Null))
}

#[cfg(test)]
mod tests {
    use super::*;
    use ariadne::{Label, Report, ReportKind, Source};
    use chumsky::input::Stream;
    use chumsky::Parser;
    use logos::Logos;

    // Macro to simplify snapshot testing for parse results
    macro_rules! assert_parse_snapshot {
        ($input:expr) => {
            assert_parse_snapshot!($input, stringify!($input))
        };
        ($input:expr, $name:expr) => {
            let result = parse_str($input);
            match result {
                Ok(stmt) => {
                    insta::assert_debug_snapshot!($name, stmt);
                }
                Err(errs) => {
                    for (i, err) in errs.iter().enumerate() {
                        let snapshot_name = if errs.len() == 1 {
                            format!("{}_error", $name)
                        } else {
                            format!("{}_error_{}", $name, i)
                        };
                        insta::assert_snapshot!(snapshot_name, err);
                    }
                }
            }
        };
    }

    // Helper function to parse a string input
    fn parse_str(input: &str) -> Result<Statement, Vec<String>> {
        let token_iter = TokenType::lexer(input)
            .spanned()
            .map(|(tok, span)| match tok {
                Ok(tok) => (tok, span.into()),
                Err(()) => (TokenType::Error, span.into()),
            });

        let token_stream =
            Stream::from_iter(token_iter).map((0..input.len()).into(), |(t, s): (_, _)| (t, s));

        parser()
            .parse(token_stream)
            .into_result()
            .map_err(|errs| build_error_message(input, errs))
    }

    fn build_error_message(source: &str, errs: Vec<Rich<TokenType, SimpleSpan>>) -> Vec<String> {
        let mut reports = Vec::new();
        for err in errs {
            let mut write_buffer = Vec::new();
            Report::build(ReportKind::Error, ((), err.span().into_range()))
                .with_config(
                    ariadne::Config::new()
                        .with_index_type(ariadne::IndexType::Byte)
                        .with_color(false),
                )
                .with_code(3)
                .with_message(err.to_string())
                .with_label(
                    Label::new(((), err.span().into_range()))
                        .with_message(err.reason().to_string()),
                )
                .finish()
                .write(Source::from(source), &mut write_buffer)
                .unwrap();
            let report = String::from_utf8_lossy(&write_buffer).to_string();
            reports.push(report);
        }
        reports
    }

    #[test]
    fn test_simple_let_declaration() {
        assert_parse_snapshot!("let x = 3;", "simple_let");
    }
}
