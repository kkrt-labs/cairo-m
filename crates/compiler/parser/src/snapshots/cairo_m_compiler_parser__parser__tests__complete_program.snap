---
source: crates/compiler/parser/src/parser.rs
expression: snapshot_entry
---
code: 
                struct Vector {
                    x: felt,
                    y: felt
                }

                namespace MathUtils {
                    func magnitude(v: Vector) -> felt {
                        return (v.x * v.x + v.y * v.y);
                    }

                    func rfib(n: felt) -> felt {
                        if (n == 0) {
                            return 0;
                        }
                        if (n == 1) {
                            return 1;
                        }
                        return rfib(n - 1) + rfib(n - 2);
                    }
                }

                const TOP_LEVEL_CONST = 100;
            
result: [
    Struct(
        Spanned(
            StructDef {
                name: Spanned(
                    "Vector",
                    24..30,
                ),
                fields: [
                    (
                        Spanned(
                            "x",
                            53..54,
                        ),
                        Named(
                            "felt",
                        ),
                    ),
                    (
                        Spanned(
                            "y",
                            82..83,
                        ),
                        Named(
                            "felt",
                        ),
                    ),
                ],
            },
            17..107,
        ),
    ),
    Namespace(
        Spanned(
            Namespace {
                name: Spanned(
                    "MathUtils",
                    135..144,
                ),
                body: [
                    Function(
                        Spanned(
                            FunctionDef {
                                name: Spanned(
                                    "magnitude",
                                    172..181,
                                ),
                                params: [
                                    Parameter {
                                        name: Spanned(
                                            "v",
                                            182..183,
                                        ),
                                        type_expr: Named(
                                            "Vector",
                                        ),
                                    },
                                ],
                                return_type: Some(
                                    Named(
                                        "felt",
                                    ),
                                ),
                                body: [
                                    Spanned(
                                        Return {
                                            value: Some(
                                                Spanned(
                                                    BinaryOp {
                                                        op: Add,
                                                        left: Spanned(
                                                            BinaryOp {
                                                                op: Mul,
                                                                left: Spanned(
                                                                    MemberAccess {
                                                                        object: Spanned(
                                                                            Identifier(
                                                                                Spanned(
                                                                                    "v",
                                                                                    235..236,
                                                                                ),
                                                                            ),
                                                                            235..236,
                                                                        ),
                                                                        field: Spanned(
                                                                            "x",
                                                                            237..238,
                                                                        ),
                                                                    },
                                                                    235..238,
                                                                ),
                                                                right: Spanned(
                                                                    MemberAccess {
                                                                        object: Spanned(
                                                                            Identifier(
                                                                                Spanned(
                                                                                    "v",
                                                                                    241..242,
                                                                                ),
                                                                            ),
                                                                            241..242,
                                                                        ),
                                                                        field: Spanned(
                                                                            "x",
                                                                            243..244,
                                                                        ),
                                                                    },
                                                                    241..244,
                                                                ),
                                                            },
                                                            235..244,
                                                        ),
                                                        right: Spanned(
                                                            BinaryOp {
                                                                op: Mul,
                                                                left: Spanned(
                                                                    MemberAccess {
                                                                        object: Spanned(
                                                                            Identifier(
                                                                                Spanned(
                                                                                    "v",
                                                                                    247..248,
                                                                                ),
                                                                            ),
                                                                            247..248,
                                                                        ),
                                                                        field: Spanned(
                                                                            "y",
                                                                            249..250,
                                                                        ),
                                                                    },
                                                                    247..250,
                                                                ),
                                                                right: Spanned(
                                                                    MemberAccess {
                                                                        object: Spanned(
                                                                            Identifier(
                                                                                Spanned(
                                                                                    "v",
                                                                                    253..254,
                                                                                ),
                                                                            ),
                                                                            253..254,
                                                                        ),
                                                                        field: Spanned(
                                                                            "y",
                                                                            255..256,
                                                                        ),
                                                                    },
                                                                    253..256,
                                                                ),
                                                            },
                                                            247..256,
                                                        ),
                                                    },
                                                    234..257,
                                                ),
                                            ),
                                        },
                                        227..258,
                                    ),
                                ],
                            },
                            167..280,
                        ),
                    ),
                    Function(
                        Spanned(
                            FunctionDef {
                                name: Spanned(
                                    "rfib",
                                    307..311,
                                ),
                                params: [
                                    Parameter {
                                        name: Spanned(
                                            "n",
                                            312..313,
                                        ),
                                        type_expr: Named(
                                            "felt",
                                        ),
                                    },
                                ],
                                return_type: Some(
                                    Named(
                                        "felt",
                                    ),
                                ),
                                body: [
                                    Spanned(
                                        If {
                                            condition: Spanned(
                                                BinaryOp {
                                                    op: Eq,
                                                    left: Spanned(
                                                        Identifier(
                                                            Spanned(
                                                                "n",
                                                                359..360,
                                                            ),
                                                        ),
                                                        359..360,
                                                    ),
                                                    right: Spanned(
                                                        Literal(
                                                            0,
                                                        ),
                                                        364..365,
                                                    ),
                                                },
                                                359..365,
                                            ),
                                            then_block: Spanned(
                                                Block(
                                                    [
                                                        Spanned(
                                                            Return {
                                                                value: Some(
                                                                    Spanned(
                                                                        Literal(
                                                                            0,
                                                                        ),
                                                                        404..405,
                                                                    ),
                                                                ),
                                                            },
                                                            397..406,
                                                        ),
                                                    ],
                                                ),
                                                367..432,
                                            ),
                                            else_block: None,
                                        },
                                        355..432,
                                    ),
                                    Spanned(
                                        If {
                                            condition: Spanned(
                                                BinaryOp {
                                                    op: Eq,
                                                    left: Spanned(
                                                        Identifier(
                                                            Spanned(
                                                                "n",
                                                                461..462,
                                                            ),
                                                        ),
                                                        461..462,
                                                    ),
                                                    right: Spanned(
                                                        Literal(
                                                            1,
                                                        ),
                                                        466..467,
                                                    ),
                                                },
                                                461..467,
                                            ),
                                            then_block: Spanned(
                                                Block(
                                                    [
                                                        Spanned(
                                                            Return {
                                                                value: Some(
                                                                    Spanned(
                                                                        Literal(
                                                                            1,
                                                                        ),
                                                                        506..507,
                                                                    ),
                                                                ),
                                                            },
                                                            499..508,
                                                        ),
                                                    ],
                                                ),
                                                469..534,
                                            ),
                                            else_block: None,
                                        },
                                        457..534,
                                    ),
                                    Spanned(
                                        Return {
                                            value: Some(
                                                Spanned(
                                                    BinaryOp {
                                                        op: Add,
                                                        left: Spanned(
                                                            FunctionCall {
                                                                callee: Spanned(
                                                                    Identifier(
                                                                        Spanned(
                                                                            "rfib",
                                                                            566..570,
                                                                        ),
                                                                    ),
                                                                    566..570,
                                                                ),
                                                                args: [
                                                                    Spanned(
                                                                        BinaryOp {
                                                                            op: Sub,
                                                                            left: Spanned(
                                                                                Identifier(
                                                                                    Spanned(
                                                                                        "n",
                                                                                        571..572,
                                                                                    ),
                                                                                ),
                                                                                571..572,
                                                                            ),
                                                                            right: Spanned(
                                                                                Literal(
                                                                                    1,
                                                                                ),
                                                                                575..576,
                                                                            ),
                                                                        },
                                                                        571..576,
                                                                    ),
                                                                ],
                                                            },
                                                            566..576,
                                                        ),
                                                        right: Spanned(
                                                            FunctionCall {
                                                                callee: Spanned(
                                                                    Identifier(
                                                                        Spanned(
                                                                            "rfib",
                                                                            580..584,
                                                                        ),
                                                                    ),
                                                                    580..584,
                                                                ),
                                                                args: [
                                                                    Spanned(
                                                                        BinaryOp {
                                                                            op: Sub,
                                                                            left: Spanned(
                                                                                Identifier(
                                                                                    Spanned(
                                                                                        "n",
                                                                                        585..586,
                                                                                    ),
                                                                                ),
                                                                                585..586,
                                                                            ),
                                                                            right: Spanned(
                                                                                Literal(
                                                                                    2,
                                                                                ),
                                                                                589..590,
                                                                            ),
                                                                        },
                                                                        585..590,
                                                                    ),
                                                                ],
                                                            },
                                                            580..590,
                                                        ),
                                                    },
                                                    566..590,
                                                ),
                                            ),
                                        },
                                        559..592,
                                    ),
                                ],
                            },
                            302..614,
                        ),
                    ),
                ],
            },
            125..632,
        ),
    ),
    Const(
        Spanned(
            ConstDef {
                name: Spanned(
                    "TOP_LEVEL_CONST",
                    656..671,
                ),
                value: Spanned(
                    Literal(
                        100,
                    ),
                    674..677,
                ),
            },
            650..678,
        ),
    ),
]
